src/compiler/tracing.ts

```diff
@@ -9,6 +9,10 @@ namespace ts.tracing {
     let legendPath: string | undefined;
     const legend: TraceRecord[] = [];
 
+    // The actual constraint is that JSON.stringify be able to serialize it without throwing.
+    type Arg = string | number | undefined;
+    type Args = Record<string, Arg | readonly Arg[]>;

```

Comment:
Maybe

    type Args = {[key: string]: string | number | Args | readonly Args[] } | undefined;

?

Also, I don't like `Args` as a name, even if it's just here...  Maybe `EventArgs`?

Reply:
1. Happy to change the name.
2. Didn't seem essential to make the type recursive yet, but I can if you feel strongly.
3. I put the `undefined` in the property value type because some of the properties are optional.

Reply:
```ts
type Args = { [key: string]: string | number | boolean | number | undefined | Args | readonly (string | number | boolean | number | undefined | Args)[] };
```

Reply:
Oh, right...  Is there some way to do something like `readonly Args[string][]` to avoid the repeating?  (If not then maybe stick with two types to avoid the repeat?)

Also, there should be a toplevel `| undefined` since they're optional, or maybe `?` the function arguments is better?

Reply:
Adopted Ron's type.